File: Data Structures & Algorithms/partition-equal-subset-sum/submission-11.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:

        s = sum(nums)
        if s%2:
            return False
        half = s//2
        N = len(nums)
        dp = {}

        def dfs(i,cur):
            print(i,cur)
            if (i,cur) in dp:
                return dp[(i,cur)]
            if cur == half:
                dp[(i,cur)] = True
                return True
            if cur>half or i == N:
                dp[(i,cur)] = False
                return False
            
            dp[(i,cur)] = dfs(i+1,cur+nums[i]) or dfs(i+1,cur)
            return dp[(i,cur)]
        return dfs(0,0)
        



        

File: Data Structures & Algorithms/burst-balloons/submission-11.java
Integer[][] dp;
class Solution {
    public int maxCoins(int[] nums) {

        int[] copy = new int[nums.length+2];
        dp = new Integer[copy.length][copy.length];
        copy[0] = copy[nums.length+1] = 1;
        int k= 1;
        for(int i =0;i<nums.length;i++){
            copy[k++]=nums[i];
        }
        return dfs(0,copy.length-1,copy);
        
    }
  
    public int dfs(int l,int r,int[] nums){
        if(r-l==1)
            return 0;
        if(dp[l][r]!=null){
            return dp[l][r];
        }
        dp[l][r] = 0;
        for(int i = l+1;i<r;i++){
            int sum = nums[l]*nums[i]*nums[r];
            dp[l][r] = Math.max(dp[l][r],sum+dfs(l,i,nums)+dfs(i,r,nums));
        }
        return dp[l][r];

    }

}
